library.ml

```diff
@@ -3,64 +3,77 @@ open Printf
 open String
 open List
 
-class enemy=object (self)
-  val mutable name:string=""
-  val mutable health:int=0
-  val mutable dmg:int=0
-  method set_health n=health <- n
-  method set_name n=name <- n
-  method set_dmg n=dmg <- n
+module Enemy = struct
+  type t =
+    { name: string ;
+      health: int ;
+      dmg: int ; }
 
-  method appear:unit=
-    printf "An enemy ";print_string [red] name;printf " appeared!\n";
-    self#battle
-  method battle:unit=
-    if health>0 then begin
-      print_string [red] name;printf " attacks you for ";
-      print_string [red] (string_of_int dmg);printf " damage\n";
+  let empty = { name = "" ; health = 0 ; dmg = 0 ; }
+
+  let rec battle t : t=
+    if t.health > 0 then begin
+      print_string [red] t.name;printf " attacks you for ";
+      print_string [red] (string_of_int t.dmg);printf " damage\n";
       printf "What will you do? (";print_string [yellow] "attack";printf ")\n";
-      let response=read_line() in begin
-        printf "\n";
-        if response="attack" then (
-          health <- health+(-1);
-          printf "You did 1 damage!\n";
-          print_string [red] name;
-          if health=0 then printf " is dead!\n" else printf " has %i health left\n" health
-        ) else printf "Your attack missed\n"
-      end;
-      self#battle
```

Comment:
Instead of `self` we have `t`. It is kind of the same, the difference is that when using `class`es you always have access to the mutable state of your objects (`self`); with modules you can write purely functional code where the state mutations are explicit.

So instead of `self#battle` we do `battle t`, which will return a type `t` (the mutated state) :)